### PR TITLE
loadgen supports multi_get_transactions

### PR DESCRIPTION
**Pull request by wlmyng**
## Description 

Describe the changes or additions included in this PR.

## Test Plan 

How did you test the new or updated feature?

---
If your changes are not user-facing and not a breaking change, you can skip the following section. Otherwise, please indicate what changed, and then add to the Release Notes section as highlighted during the release process.

### Type of Change (Check all that apply)

- [ ] user-visible impact
- [ ] breaking change for a client SDKs
- [ ] breaking change for FNs (FN binary must upgrade)
- [ ] breaking change for validators or node operators (must upgrade binaries)
- [ ] breaking change for on-chain data layout
- [ ] necessitate either a data wipe or data migration

### CODE REVIEW

File: crates/sui-indexer/src/lib.rs [hunk 1]
```diff
@@ -176,7 +176,7 @@ pub fn establish_connection(db_url: String) -> PgConnection {
 pub async fn new_pg_connection_pool(db_url: &str) -> Result<PgConnectionPool, IndexerError> {
     let manager = ConnectionManager::<PgConnection>::new(db_url);
     // default connection pool max size is 10
-    Pool::builder().build(manager).map_err(|e| {
+    Pool::builder().max_size(700).build(manager).map_err(|e| {
```
Comment (666lcz) [inline crates/sui-indexer/src/lib.rs:179]: looks like you are based on a older commit

File: crates/sui-rpc-loadgen/README.md [hunk 2]
```diff
@@ -16,6 +16,8 @@ Run the following command to see available commands:
 cargo run --bin sui-rpc-loadgen -- -h
 ```
 
+To try this locally, refer to [sef](../sui-test-validator/README.md). Recommend setting `database-url` to an env variable.
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/README.md:19]: 👍

File: crates/sui-rpc-loadgen/README.md [hunk 3]
```diff
@@ -35,4 +37,16 @@ Both commands achieve the same outcome: fetching all checkpoints using one threa
...
\ No newline at end of file
+**NOTE**: right now `pay-sui` only supports 1 thread but multi-threading support can be added pretty easily by assigning different gas coins to different threads
+
+
+### Example 3: Get All Transactions
+Finds all checkpoints, and then all transactions in those checkpoints
+
+
+## Adding other endpoints
+
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/README.md:48]: Thanks for adding this! After your PR is merged, we can also add link to this PR as an example

File: crates/sui-rpc-loadgen/src/main.rs [hunk 4]
```diff
@@ -69,6 +70,21 @@ pub enum ClapCommand {
...
         common: CommonOptions,
     },
+    #[clap(name = "multi-get-transactions")]
+    MultiGetTransactions {
+        #[clap(flatten)]
+        common: CommonOptions,
+        // this is where the vector goes
+        /// Default to start from checkpoint 0
+        #[clap(short, long, default_value_t = 0)]
+        start: u64,
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/src/main.rs:80]:
```suggestion
        start_checkpoint: u64,
```
nit:

File: crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs [hunk 5]
```diff
@@ -211,10 +211,110 @@ impl<'a> ProcessPayload<'a, &'a GetCheckpoints> for RpcCommandProcessor {
...
-        _op: &'a MultiGetTransactions,
+        op: &'a MultiGetTransactions,
         _keypair: &Option<SuiKeyPair>,
     ) -> Result<()> {
-        debug!("MultiGetTransactions");
+
+        // if digests provided
+        if op.digests.is_some() {
+
+        } else {
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs:221]: Feel free to add a TODO and  tackle this as a separate PR, we don't need this right now.

File: crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs [hunk 6]
```diff
@@ -211,14 +212,168 @@ impl<'a> ProcessPayload<'a, &'a GetCheckpoints> for RpcCommandProcessor {
...
+    let read = clients.read().await;
+    let clients = read.clone();
+
+    let transactions = join_all(clients.iter().enumerate().map(|(i, client)| async move {
+        let start_time = Instant::now();
+        let transactions = client
+            .read_api()
+            .multi_get_transactions_with_options(
+                digests.to_vec(),
+                SuiTransactionResponseOptions::new(),
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs:339]: Ah, this by default is only going to return `digest`. You need to specify some options, let's start with input and effects

File: crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs [hunk 7]
```diff
@@ -211,14 +212,168 @@ impl<'a> ProcessPayload<'a, &'a GetCheckpoints> for RpcCommandProcessor {
...
+            elapsed_time.as_secs_f64()
+        );
+        transactions
+    }))
+    .await;
+
+    if transactions.len() == 2 {
+        if let (Some(t1), Some(t2)) = (transactions.get(0), transactions.get(1)) {
+            let first = t1.as_ref().unwrap();
+            let second = t2.as_ref().unwrap();
```
Comment (666lcz) [inline crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs:354]: use expect() with object_id info to help debug

File: crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs [hunk 8]
```diff
@@ -211,14 +212,168 @@ impl<'a> ProcessPayload<'a, &'a GetCheckpoints> for RpcCommandProcessor {
...
+        let start_time = Instant::now();
+        let transactions = client
+            .read_api()
+            .multi_get_transactions_with_options(
+                digests.to_vec(),
+                SuiTransactionResponseOptions::new(),
+            )
+            .await;
+        let elapsed_time = start_time.elapsed();
+        println!(
```
[the remark this answers is not in the document]
Comment (wlmyng) [reply crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs:343]: sure thing

File: crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs [hunk 9]
```diff
@@ -211,14 +212,168 @@ impl<'a> ProcessPayload<'a, &'a GetCheckpoints> for RpcCommandProcessor {
...
+            "MultiGetTransactions Request latency {:.4} for rpc at url {i}",
+            elapsed_time.as_secs_f64()
+        );
+        transactions
+    }))
+    .await;
+
+    if transactions.len() == 2 {
```
[the remark this answers is not in the document]
Comment (wlmyng) [reply crates/sui-rpc-loadgen/src/payload/rpc_command_processor.rs:351]: error makes sense

**Pull request closed without merging**